### PR TITLE
UHV: Transform H/2 extended CONNECT to H/1 upgrade

### PR DESCRIPTION
**Pull request by yanavlasov**
Additionally:
* split header validator interface for server and client validation as the behavior is different
* Plumb UHV into test client and fake upstream
 
Risk Level: Low, build flag protected
Testing: Unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Part of #24945

### CODE REVIEW

File: envoy/http/header_validator.h
Comment (alyssawilk) [inline envoy/http/header_validator.h:153]: stepping back a bit I want to have a high level discussion about validation vs normalization.
It's surely more efficient to do the two together but from an API perspective feels a lot cleaner to do it in 2 steps.
Also don't we do more than just connect based transforms?  I thought the whole drop_headers_with_underscore path would need new headers as well?
/wait-any

File: source/common/http/codec_client.cc
Comment (alyssawilk) [inline source/common/http/codec_client.cc:195]: Is this ready for review or do you want to do a clean up pass.
Also I tend to not do major refactors along with functional change so I'd mildly lean towards another refactor PR followed by the change, but as this code isn't in prod I'll leave it to your call
/wait on clean-up

Comment (yanavlasov) [reply source/common/http/codec_client.cc]: Should be good fore review now.

File: source/extensions/http/header_validators/envoy_default/http2_header_validator.cc
Comment (alyssawilk) [inline source/extensions/http/header_validators/envoy_default/http2_header_validator.cc:441]: I really don''t love that both paths return a transformation result with potential new headers but this changes headers inline.  I feel like it's a footgun waiting to happen.  I'd rather fix the APIs to be consistent
cc @mattklein123 for thoughts

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo minor nits.  Over to Matt for !google pass and second opinion

File: source/extensions/http/header_validators/envoy_default/http2_header_validator.cc
Comment (yanavlasov) [reply source/extensions/http/header_validators/envoy_default/http2_header_validator.cc:441]: The main issue here is not so much with the API but with the violation of invariant that header map is not mutated by the upstream codec. Violation of this invariant breaks a number of places. For example, HCM encoder path breaks because it observed an upgrade request on the decode path and then extended CONNECT request when processing response. This also breaks retries and hedging to a mixed protocol pool, which now needs to transform  in both H/1 and H/2 codecs.

Changing it is doable but is a significant refactor. It is also unclear if there could any effect on filters that may keep state based on the type of request, which is changed when it is sent to the upstream.

We can tackle it, but it is a dangerous change.

**Comment by yanavlasov** [comment]
Fixing coverage.

File: source/extensions/http/header_validators/envoy_default/http2_header_validator.cc
Comment (mattklein123) [reply source/extensions/http/header_validators/envoy_default/http2_header_validator.cc:441]: I honestly don't have any context on this ongoing work to understand the concern. If we aren't going to fix Alyssa's concerns now can we just add more comments/TODO to clearly spell them out? Thank you.

Comment (yanavlasov) [reply source/extensions/http/header_validators/envoy_default/http2_header_validator.cc:441]: yes, i think we can do that. Let me craft the PR to check this.

Comment (alyssawilk) [reply source/extensions/http/header_validators/envoy_default/http2_header_validator.cc:441]: ah, I think I was conflating RequestHeaderValidationResult with ResponseHeaderValidationResult.

sigh.
OK no objections to the API then, will review

**Review by alyssawilk — APPROVED** [review]

**Comment by phlax** [comment]
this fails clang-tidy

```diff
diff --git a/test/common/http/codec_client_test.cc b/test/common/http/codec_client_test.cc
index ffb60f2a19..3063e35e1e 100644
--- a/test/common/http/codec_client_test.cc
+++ b/test/common/http/codec_client_test.cc
@@ -30,7 +30,6 @@
 
 using testing::_;
 using testing::AtMost;
-using testing::ByMove;
 using testing::Invoke;
 using testing::InvokeWithoutArgs;
 using testing::NiceMock;

```

https://dev.azure.com/cncf/envoy/_build/results?buildId=135536&view=logs&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390&l=253

**Pull request merged**
